### ISSUE TITLE
Users reported not been able to login to Pixelated

### ISSUE THREAD

**Issue by deniscostadsc** [pixelated/pixelated-user-agent#795]
**Problem description:**

Users reported not been able to login to Pixelated. They had to try a second time to access their inboxes.

**Definition of done:**

Find out what make them unable to login and fix it.

**Comment by deniscostadsc** [comment]
We are waiting for the logs.

**Comment by deniscostadsc** [comment]
I've found out that most of user are adding the domain to their username when they try to login. Pixelated doesn't support login with domain.

**Comment by deniscostadsc** [comment]
@bwagnerr remove text after '@'. Should we put it to review?

**Comment by rdevries-thoughtworks** [comment]
Let's coordinate the required login refactorings a bit, since #791 also seems to require some refactoring.

**Comment by deniscostadsc** [comment]
@ayoyoness

This issue is been developed in this branch https://github.com/pixelated/pixelated-user-agent/tree/refactor_auth_out_of_resource.

I split into two different classes just to keep the old API. We can change those classes when we start to pair.

**Comment by deniscostadsc** [comment]
I rewrote the commit messages. So I did some squash. and push force.

**Comment by deniscostadsc** [comment]
We finished this task.

We did a lot of refactoring in order to be able to test the authentication process. Auth is something critical regarding security, so please take a look carefully when reviewing this issue.

**Comment by thaissiqueira** [comment]
@deniscostadsc and @ayoyoness 

About security, there are some specific thing that we only should do on review phase?
I would like to understand better your point because now the code are already on master right?
I wondering if this care don't should be taken during the development phase. 
There are some thing that we can do on development phase to prevent security issues?

**Comment by ayoyoness** [comment]
oi
please add more lines (before and after) on the error  log.
Thanks,

**Comment by thaissiqueira** [comment]
Oi!

Follow the whole log bellow:

```
2016-11-01 12:06:34 [leap.bitmask.bonafide.session] DEBUG POST to https://api.staging.pixelated-project.org:4430/1/sessions
2016-11-01 12:06:34 [twisted.web.client._HTTP11ClientFactory] INFO Starting factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f3709158440>
2016-11-01 12:06:35 [leap.bitmask.bonafide.session] DEBUG PUT to https://api.staging.pixelated-project.org:4430/1/sessions/thais
2016-11-01 12:06:35 [twisted.web.client._HTTP11ClientFactory] INFO Starting factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f37080f0680>
2016-11-01 12:06:35 [twisted.web.client._HTTP11ClientFactory] INFO Stopping factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f3709158440>
2016-11-01 12:06:37 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:36 +0000] "POST /login HTTP/1.1" 200 624 "http://localhost:3333/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:37 [leap.soledad.client.secrets] INFO trying to load secrets from local storage...
2016-11-01 12:06:37 [twisted.web.client._HTTP11ClientFactory] INFO Stopping factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f37080f0680>
2016-11-01 12:06:37 [leap.soledad.client.secrets] INFO found secrets in local storage
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.common.events.client] DEBUG Subscribing to event: KEYMANAGER_FINISHED_KEY_GENERATION
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.common.http._HTTP11ClientFactory] INFO Starting factory <leap.common.http._HTTP11ClientFactory instance at 0x7f37080627a0>
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 12:06:37 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 12:06:37 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:36 +0000] "GET /favicon.ico HTTP/1.1" 401 13 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG target replica uid: 0ba4d2145c7d4c7fbc4dc0bfaaaba725
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG target generation: 14
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG target trans id: T-048d3cac3b47446fa11890a84c50aefd
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG target my gen: 14
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG target my trans_id: T-33a7838fad7a4c91bd0795116f82dd17
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG source replica_uid: 6a156ae8e9db41caa58c567104c4b1b4
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG there are 0 documents to send
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG last known target gen: 14
2016-11-01 12:06:38 [leap.soledad.client.sync] DEBUG last known target trans_id: T-048d3cac3b47446fa11890a84c50aefd
2016-11-01 12:06:38 [leap.common.events.client] DEBUG Creating client instance.
2016-11-01 12:06:38 [leap.common.events.client] DEBUG Emitting event: (SOLEDAD_DONE_DATA_SYNC, ({'userid': u'e72894f4bdd2f032163b43dff15e1577', 'uuid': u'e72894f4bdd2f032163b43dff15e1577'},))
2016-11-01 12:06:38 [leap.bitmask.keymanager] DEBUG getting key for thais@staging.pixelated-project.org
2016-11-01 12:06:38 [leap.common.events.client] DEBUG Emitting event: (KEYMANAGER_LOOKING_FOR_KEY, (u'thais@staging.pixelated-project.org',))
2016-11-01 12:06:38 [leap.common.events.client] DEBUG Emitting event: (KEYMANAGER_KEY_NOT_FOUND, (u'thais@staging.pixelated-project.org',))
2016-11-01 12:06:38 [pixelated.resources.login_resource] ERROR Login error during thais services setup: thais@staging.pixelated-project.org
2016-11-01 12:06:38 [leap.common.events.server] DEBUG Publishing event: SOLEDAD_DONE_DATA_SYNC
2016-11-01 12:06:38 [leap.common.events.server] DEBUG Publishing event: KEYMANAGER_LOOKING_FOR_KEY
2016-11-01 12:06:38 [leap.common.events.server] DEBUG Publishing event: KEYMANAGER_KEY_NOT_FOUND
2016-11-01 12:06:39 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:38 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:41 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:40 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:43 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:42 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:45 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:44 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 12:06:47 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:12:06:46 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
^C2016-11-01 12:06:48 [log_legacy] INFO Received SIGINT, shutting down.
2016-11-01 12:06:48 [leap.common.http._HTTP11ClientFactory] INFO Stopping factory <leap.common.http._HTTP11ClientFactory instance at 0x7f37080627a0>
2016-11-01 12:06:48 [log_legacy] INFO (TCP Port 3333 Closed)
2016-11-01 12:06:48 [pixelated.config.site.PixelatedSite] INFO Stopping factory <pixelated.config.site.PixelatedSite instance at 0x7f3709db8c68>
2016-11-01 12:06:48 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:48 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:48 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:48 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:48 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 12:06:48 [log_legacy] INFO Main loop terminated.
```

**Comment by thaissiqueira** [comment]
@ayoyoness 

I putted 'error.value()' in [login_resource.py#L168](https://github.com/pixelated/pixelated-user-agent/blob/master/service/pixelated/resources/login_resource.py#L168) line in order to got more informations about the error and the result was:

```
2016-11-01 18:28:27 [leap.bitmask.bonafide.session] DEBUG POST to https://api.staging.pixelated-project.org:4430/1/sessions
2016-11-01 18:28:27 [twisted.web.client._HTTP11ClientFactory] INFO Starting factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f2e8c0af488>
2016-11-01 18:28:29 [leap.bitmask.bonafide.session] DEBUG PUT to https://api.staging.pixelated-project.org:4430/1/sessions/thais
2016-11-01 18:28:29 [twisted.web.client._HTTP11ClientFactory] INFO Starting factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f2e8c0aaf80>
2016-11-01 18:28:29 [twisted.web.client._HTTP11ClientFactory] INFO Stopping factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f2e8c0af488>
2016-11-01 18:28:30 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:30 +0000] "POST /login HTTP/1.1" 200 624 "http://localhost:3333/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:30 [leap.soledad.client.secrets] INFO trying to load secrets from local storage...
2016-11-01 18:28:30 [twisted.web.client._HTTP11ClientFactory] INFO Stopping factory <twisted.web.client._HTTP11ClientFactory instance at 0x7f2e8c0aaf80>
2016-11-01 18:28:30 [leap.soledad.client.secrets] INFO found secrets in local storage
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 18:28:30 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting write-ahead logging
2016-11-01 18:28:30 [leap.common.events.client] DEBUG Subscribing to event: KEYMANAGER_FINISHED_KEY_GENERATION
2016-11-01 18:28:30 [leap.soledad.client.pragmas] DEBUG sqlcipher: setting synchronous normal
2016-11-01 18:28:30 [leap.common.http._HTTP11ClientFactory] INFO Starting factory <leap.common.http._HTTP11ClientFactory instance at 0x7f2e8c01e7e8>
2016-11-01 18:28:30 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:30 +0000] "GET /favicon.ico HTTP/1.1" 401 13 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG target replica uid: 0ba4d2145c7d4c7fbc4dc0bfaaaba725
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG target generation: 14
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG target trans id: T-048d3cac3b47446fa11890a84c50aefd
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG target my gen: 10
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG target my trans_id: T-0b818fb95596440486fb950d89f26e31
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG source replica_uid: 36a31d670f47411682fc56523bb81450
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG there are 0 documents to send
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG last known target gen: 14
2016-11-01 18:28:31 [leap.soledad.client.sync] DEBUG last known target trans_id: T-048d3cac3b47446fa11890a84c50aefd
2016-11-01 18:28:31 [leap.common.events.client] DEBUG Creating client instance.
2016-11-01 18:28:31 [leap.common.events.client] DEBUG Emitting event: (SOLEDAD_DONE_DATA_SYNC, ({'userid': u'e72894f4bdd2f032163b43dff15e1577', 'uuid': u'e72894f4bdd2f032163b43dff15e1577'},))
2016-11-01 18:28:31 [leap.bitmask.keymanager] DEBUG getting key for thais@staging.pixelated-project.org
2016-11-01 18:28:31 [leap.common.events.client] DEBUG Emitting event: (KEYMANAGER_LOOKING_FOR_KEY, (u'thais@staging.pixelated-project.org',))
2016-11-01 18:28:31 [leap.common.events.client] DEBUG Emitting event: (KEYMANAGER_KEY_NOT_FOUND, (u'thais@staging.pixelated-project.org',))
2016-11-01 18:28:31 [twisted.internet.defer] CRITICAL Unhandled error in Deferred:
2016-11-01 18:28:31 [leap.common.events.server] DEBUG Publishing event: SOLEDAD_DONE_DATA_SYNC
2016-11-01 18:28:31 [leap.common.events.server] DEBUG Publishing event: KEYMANAGER_LOOKING_FOR_KEY
2016-11-01 18:28:31 [leap.common.events.server] DEBUG Publishing event: KEYMANAGER_KEY_NOT_FOUND
2016-11-01 18:28:32 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:32 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:34 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:34 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:36 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:36 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:38 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:38 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
2016-11-01 18:28:40 [twisted.python.log] INFO "10.0.2.2" - - [01/Nov/2016:18:28:40 +0000] "GET / HTTP/1.1" 200 1441 "http://localhost:3333/login" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36"
^C2016-11-01 18:28:41 [log_legacy] INFO Received SIGINT, shutting down.
2016-11-01 18:28:41 [leap.common.http._HTTP11ClientFactory] INFO Stopping factory <leap.common.http._HTTP11ClientFactory instance at 0x7f2e8c01e7e8>
2016-11-01 18:28:41 [log_legacy] INFO (TCP Port 3333 Closed)
2016-11-01 18:28:41 [pixelated.config.site.PixelatedSite] INFO Stopping factory <pixelated.config.site.PixelatedSite instance at 0x7f2e8d573cb0>
2016-11-01 18:28:41 [leap.soledad.client.sqlcipher] DEBUG SQLCipher backend: closing
2016-11-01 18:28:41 [log_legacy] INFO Main loop terminated.
2016-11-01 18:28:41 [twisted.internet.defer] CRITICAL Unhandled error in Deferred:
```

**Comment by thaissiqueira** [comment]
@ayoyoness 

My fault! I left some old code there. 
Now it is working! thx!

**Comment by rdevries-thoughtworks** [comment]
I found this diff in commit 423ca8f9fb7636b336b24ba28bde5d61538bf5fc:

```diff
-    try:
-        auth = yield authenticate(provider, username, password)
-    except SRPAuthError:
-        raise UnauthorizedLogin()
+    auth = Authenticator(provider).authenticate(username, password)
```

It seems like the result of the `authenticate` method should be yielded. But more importantly, we should have a test that catches this.

**Comment by ayoyoness** [comment]
@rdevries-thoughtworks 
That overall method actually orchestrates the overall leap initialization for the single user mode.
It is a bit hard to do a meaningful test on it as it integrates most of the components of the user-agent, that includes file I/O, http calls, and access to soledad/keymanager... the functional tests as end-to-end should reveal the bugs. In this particular case, it is still green because, actually the `create_leap_session` method that comes up after the authentication part has a fallback to re-authenticate if the user hasn't yet... so it has not failed.
So overall, no loss in functionality but in performance and consistency.
I added a unit test for it now.

**Comment by rdevries-thoughtworks** [comment]
After discussing with @ayoyoness:
- there actually doesn't seem to be a auth-fallback in the `create_leap_session`
- it's not clear (at least to me) why that fallback was there for in the first place
- the functional tests are not failing, but the user agent in single user mode does not start
- there should be a functional test for at least the single user mode and multi-user mode that verifies the refactored code is working

@bwagner: including you because you asked :-)

**Comment by bwagnerr** [comment]
@rdevries-thoughtworks the current functional tests don't test one bootstrap (multi-user) or another (single-user), they test the functionality considering there's a user logged in. We have some rough separation of multi and single user on the integration tests only.

About the functional tests we should add, it would be good to incorporate that to the work on making the user agent functional tests into end-to-end tests, I don't think it is worth spending the effort on that now

**Comment by rdevries-thoughtworks** [comment]
> About the functional tests we should add, it would be good to incorporate that to the work on making the user agent functional tests into end-to-end tests, I don't think it is worth spending the effort on that now

I see how it would fit in to the end-to-end tests issue, but I don't think we can move this issue to done without testing it. You can't put basically broken code (because untested) into production because the tests are also part of a different issue.

**Comment by bwagner** [comment]
Guys, please note that my moniker is bwagner, while your friend's is bwagnerr. I keep getting stuff I'm not associated with. Thank you!

**Comment by rdevries-thoughtworks** [comment]
@bwagner Really sorry for that :-(. I'll pay attention to it in the future.

**Comment by ayoyoness** [comment]
@thaissiqueira and I just pushed a commit that fixes the single user mode startup issue.
We did not add a test for it because the broken method is at the highest level of the application (bootstrap) and unit-testing it would just fit the test for that particular -broken- case and will not be of greater importance.
I agree with @bwagnerr that we should not spend time on trying to figure this out now as it is already part of the functional test review strategy that we have. 
At the same time I understand @rdevries-thoughtworks's point of view. I suggest we put this on blocked until we implement the functional tests review.

**Comment by thaissiqueira** [comment]
I'm trying to implement (with @bwagnerr's helps) an integration test to prevent the activist mode error that we had during this issue review.

**Comment by thaissiqueira** [comment]
As mentioned above, this implementation is working fine but we can't consider this issue done yet because is missing test (at least to prevent the single user error that was fixed here: e73c48bff2574a9678f90aa56066aec39a240e1e).
Because of that I'm putting it on ready.

**Comment by anikarni** [comment]
After talking with @thaissiqueira and @tayanefernandes , we decided to draw a strategy for the missing test, which might require a bit more planning. Therefore, I'm closing this issue.
